### PR TITLE
[CINFRA] Refactoring: Emancipate replicated logs and states

### PR DESCRIPTION
**Pull request by maierlars**
# Scope & Purpose

In order to implement any kind of volatile replication, i.e. `waitForSync=false` or ActiveFailover, we need more complex interaction between replicated logs and replicated states. The previous structure, where the log knew nothing of the state, while the state did own the log, does not work any longer. This was already inconvenient, as the state had to react on some events in the log, which weren't directly communicated (because the log couldn't call the state), so had to rely on more indirect observations instead. This made the code harder to read (&write), and more fragile.

Instead, both are now emancipated, and mutually exchange structure for callbacks to the respective other.

Also, this moves to a merged view of both entities, with the goal that later there's only one entity from an outside user's perspective. This is currently neither complete nor completely consistent, however.

This change does not affect the API between the replicated state and the state implementation.

## Main changes

- The log and state data types are now emancipated. The state doesn't "own" the log any longer, without the log knowing anything of the state. Instead, they exchange mutual callback-structures.
  - For example, this means if the log gets in a situation where the state has to acquire a new snapshot, it can immediately mark the snapshot as being unavailable, and direct the state to get a new one. Instead of relying on the state to observe indirectly that this is necessary.
  - replicated state managers are rewritten: instead of the explicit state machine, we now use the callbacks between state and log to pass control.
- `{Target,Plan,Current}/ReplicatedStates` are removed from the agency store, there's only `{Target,Plan,Current}/ReplicatedLogs` now.
- The snapshot's state is now owned by ReplicatedLogs (instead of ~States). It is also persisted now, by the LogCore. This introduces "snapshotAvailable" for participants, and also allowed the introduction of the "kSnapshotMissing" reason for not being allowed in a Quorum.
  - The snapshot's state is also communicated by the followers to their leader. This can happen either via the newly introduced `update-snapshot-status` API (see `RestLogInternalHandler`), or in the response to an append-entries request. Races between the two are possible, but this has been taken care of via MessageIDs. The additional API is necessary, as the follower can't send a response without a previous request, which could possibly result in stagnancy.
- In RocksDB, metadata for replicated logs and their corresponding states is no longer persisted separately (there is just ReplicatedState now).
     - The Storage Engine Interfaces implementation provided by rocksdb has been refactored. It now provides the Sequence number of an operation as well. This is later used to implement a sync index.
     - Furthermore the storage engine now ensures that there is no ongoing async write operation on a replicated log before its meta data is deleted. Previously there was a race and random log slices could just stay there forever.

## Minor changes

- Introduced CompactionStopReason, indicating why the currently oldest LogIndex was not compacted (yet).
- Introduced "LeaderCommunicator" classes, which allow the followers to send network requests to the leader, in addition to just being allowed to send responses to append-entries request. This is also a first (tiny) step to move towards a full-duplex bidirectional streaming protocol for the communication between leaders and followers.
- Added --replicated-log.threshold-log-compaction (Number of log entries to wait for before compacting).
- Increased default log levels of replication2 and rep-state to WARN

## Unrelated changes

- Allowed a larger buffer size in the terminate handler for exception messages
- Made exitcodes.js available in JS, like errors.js already is

## Bugfixes

- On leader construction, the leader assumed its follower's last log entry to be one less than its own. This was done in the early development phases so an AppendEntries was triggered without an insert. This is not necessary anymore (e.g. because of the log entry introduced when establishing leadership). But more importantly, due to compaction, this could lead to the leader instructing its followers to throw away their snapshots.
- The leader calculated the `lowestIndexToKeep` of its followers incorrectly, which could lead to overeager compaction, and subsequently to follower(s) losing their snapshots. It used the `lastAckedCommitIndex` instead of the `lastAckedIndex`: The former is just the last commit index the follower has seen from the leader (just the index), while the latter is the last acknowledged entry from the follower. If enough participants are ahead, the `lastAckedCommitIndex` of some (lagging) followers can be larger than their corresponding `lastAckedIndex`.
- If a server is stopped and its reboot id was changed by the supervision, it is not updated internally. This results in the server becoming a follower for that term and never assumes leadership again.

### CODE REVIEW

**Review by MBkkt — COMMENTED** [review]
Please run sanitizers build

File: arangod/Replication2/ReplicatedState/ReplicatedState.tpp
Comment (MBkkt) [inline arangod/Replication2/ReplicatedState/ReplicatedState.tpp:219]: dynamic cast is very expensive. Do you really need it in this file?
Maybe you can use basics::downCast? It also works with shared_ptr

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: Better to don't use std::atomic::wait/notify it contains a lot of bugs right now (deadlocked in some cases)!

Also if you want to use syscall futex, ulock, waitonaddress, it should be int!
For other types it will be use small (16 in libstdc++) static table of int for wait/wake

File: arangod/RocksDBEngine/RocksDBPersistedLog.cpp
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBPersistedLog.cpp:52]:
```suggestion
    _pendingAsyncOperations.wait(c, std::memory_order_relaxed);
```

File: arangod/RocksDBEngine/RocksDBPersistedLog.cpp
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBPersistedLog.cpp:64]:
```suggestion
  _pendingAsyncOperations.fetch_add(1, std::memory_order_relaxed);
```

**Review by MBkkt — CHANGES_REQUESTED** [review]

**Comment by MBkkt** [comment]
> https://jenkins01.arangodb.biz/job/arangodb-devel-asan-and-coverage-biweekly/385/

https://jenkins01.arangodb.biz/job/arangodb-ANY-linux-san.x86-64/1478/EDITION=enterprise,SAN_MODE=AULSan,STORAGE_ENGINE=rocksdb,TEST_SUITE=single,limit=linux&&test&&x86-64/artifact/aulsan.log.arangodbtests.75980.log

Fix for this issue was merged in devel

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (mpoeter) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: Where did you get that atomic wait/notify implementations are buggy? I know that the glibc sempahore implementation is buggy, but only because the atomic wait/notify _usage_ is incorrect.

Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: https://gcc.gnu.org/bugzilla/buglist.cgi?quicksearch=atomic%20wait
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106183 fixed in trunk
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101037
maybe other

Also libstdc++ semaphore don't use atomic it use separate call with platform wait/wake (because other methods have spin wait which implemented in semaphore)

**Comment by MBkkt** [comment]
https://jenkins01.arangodb.biz/job/arangodb-ANY-linux-san.x86-64/1478/EDITION=enterprise,SAN_MODE=TSan,STORAGE_ENGINE=rocksdb,TEST_SUITE=single,limit=linux&&test&&x86-64/artifact/tsan.log.arangodbtests.70895.log

Some new data race in test

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: If I read that correctly, there are bugs in `libstdc++`'s fallback implementation, when `_GLIBCXX_HAVE_PLATFORM_WAIT` is false. CMIIW, but I think we're not affected; as this should be true for Linux, on Windows we're using Microsoft's STL, and on Mac we're using `libc++` (and maybe it also has platform wait, I don't know).

Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: @goedderz if I understand correctly you not fully right.
These bugs in "not platform atomic wait", in waiter pool, waiter pool used always when T from atomic<T> is not int (for libc++) and sizeof(int) for libstdc++.

Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: > and for std::atomic<T> types where T is different from __platform_wait_t.

I found, you probably missing this part from issue

Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]:
```cpp
#if !defined(_LIBCPP_VERSION) || defined(__linux__) || (defined(_AIX) && !defined(__64BIT__))
  using wait_t = std::int32_t;
#else
  using wait_t = std::int64_t;
#endif
```

**Review by apetenchea — APPROVED** [review]
I already had an overview from out meeting, but I went again through the files. I paid more attention to the `Replication2` directory and tried to move a bit quicker through the rest.
Apart from the few comments I left, looks good to me!

**Comment by goedderz** [comment]
> https://jenkins01.arangodb.biz/job/arangodb-ANY-linux-san.x86-64/1478/EDITION=enterprise,SAN_MODE=TSan,STORAGE_ENGINE=rocksdb,TEST_SUITE=single,limit=linux&&test&&x86-64/artifact/tsan.log.arangodbtests.70895.log
> 
> Some new data race in test

Thanks, will have a look.

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: Thanks for clearing that up @MBkkt. We should probably find out then whether the fix (https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=af98cb88eb4be6a1668ddf966e975149bf8610b1) has made it into the libstdc++ version we're using yet. Or I'll just switch to another primitive for now.

Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:49]: Unfortunately we definitely don't have such fix, it committed in this year, we use alpine with gcc 11.2 so..

But maybe int is enough for your purpose?

Anyway you are always can use condition_variable + mutex + atomic_size_t
Or absl::Mutex + atomic_size_t if you're caring about sizeof.

In general it will be works maybe better than waiters pool.

File: arangod/Cluster/UpdateReplicatedLogAction.cpp
Comment (MBkkt) [reply arangod/Cluster/UpdateReplicatedLogAction.cpp:67]: fixed?

File: lib/Basics/VelocyPackHelper.h
Comment (goedderz) [reply lib/Basics/VelocyPackHelper.h:459]: Trailing return types aren't specific to replication 2 code, if that's what you mean.

File: arangod/VocBase/vocbase.h
Comment (goedderz) [reply arangod/VocBase/vocbase.h:204]: We don't really need FlatHashMap. :)
It just doesn't matter here, so we're sticking with the STL containers for now. Changing to another container is a micro-optimization that's far from useful in the current state, and I'd rather do only when the code is much more stable (stable in the "isn't going to change much any more" sense).

File: arangod/Replication2/Methods.cpp
Comment (goedderz) [reply arangod/Replication2/Methods.cpp:806]: We're currently not doing micro optimizations while the code is still changing as much as it does.

File: arangod/Replication2/Methods.cpp
Comment (goedderz) [reply arangod/Replication2/Methods.cpp:398]: https://github.com/arangodb/arangodb/pull/17528#discussion_r1053400545

File: arangod/Replication2/Methods.cpp
Comment (goedderz) [reply arangod/Replication2/Methods.cpp:798]: https://github.com/arangodb/arangodb/pull/17528#discussion_r1053400545

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h]: > But maybe int is enough for your purpose?

Yes, I've decided to go with [your recommendation](https://github.com/arangodb/arangodb/pull/17528/files/ffd79b1ad531797fb504c401753c08c71933a3b9?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1050126579).

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:42]: While you're right that mostly C-style is still being used, the namespaced types are already being used a lot:
```
$ rg -Io '\bstd::(u?int\d+|size)_t\b' lib/ client-tools/ arangod/ | sort -n | uniq -c
      2 std::int16_t
      3 std::int32_t
     51 std::int64_t
    723 std::size_t
      2 std::uint16_t
    203 std::uint32_t
    923 std::uint64_t
     44 std::uint8_t
```
I also think there's nothing speaking against using it for new code.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.h
Comment (goedderz) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.h:50]: I've thought of that, but it's not important enough for me to introduce an `#ifdef` that might bite someone only after running Jenkins.

File: arangod/Replication2/ReplicatedState/WaitForQueue.h
Comment (goedderz) [reply arangod/Replication2/ReplicatedState/WaitForQueue.h:45]: We will probably consider a different data structure for this down the road. But we might also replace it with a different mechanism entirely. As the multimap is well encapsulated in the `WaitForQueue`, so it'll be trivial to replace it later, we're currently not investing any time.

File: arangod/Replication2/Methods.h
Comment (goedderz) [reply arangod/Replication2/Methods.h:138]: It's of no interest in this code path; so [as above](https://github.com/arangodb/arangodb/pull/17528#discussion_r1053396589), we're sticking with STL containers for now.

File: arangod/Replication2/ReplicatedState/ReplicatedState.tpp
Comment (goedderz) [reply arangod/Replication2/ReplicatedState/ReplicatedState.tpp:219]: `getFollower()` is currently not called at all in any production code, only in tests, and `getLeader()` is called only once in `ensureCollection`. I currently see no need to change the performance, and rather have the tiny safety guarantees a dynamic cast provides.
Also, this code path is subject to change, which is another reason why I currently don't want to bother with micro optimizations here.

File: arangod/RocksDBEngine/RocksDBPersistedLog.cpp
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.cpp:52]: I think you're right, but I don't want to change this without checking back with @maierlars, as he authored this part. However, as he's out for a few days, I'd like to discuss this with him after the merge, in order not to delay it further, if that's fine with you.

File: arangod/RocksDBEngine/RocksDBPersistedLog.cpp
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.cpp:64]: https://github.com/arangodb/arangodb/pull/17528#discussion_r1053434825

File: lib/Basics/VelocyPackHelper.h
Comment (MBkkt) [reply lib/Basics/VelocyPackHelper.h:459]: > Trailing return types aren't specific to replication 2 code, if that's what you mean.

I think it is specific. Other directories written in other codestyle

File: arangod/VocBase/vocbase.h
Comment (MBkkt) [reply arangod/VocBase/vocbase.h:204]: The main issue to unordered map (expect it's slower): it gives a lot of commonly unnecessary guarantees.
And in future changing it will be more difficult, because needs to understand is this guarantees necessary or not?

File: arangod/Replication2/Methods.cpp
Comment (MBkkt) [reply arangod/Replication2/Methods.cpp:806]: It's not micro optimization it's just correct way to write code:
if you make cycle and push to container is common to make reserve before it.

Even in clang-tidy exist such rule, so it's same as speak it's ok to use c-style cast here instead of static_cast

https://clang.llvm.org/extra/clang-tidy/checks/performance/inefficient-vector-operation.html

File: arangod/RocksDBEngine/RocksDBPersistedLog.h
Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.h:42]: Something like 10% or smaller, but as I wrote it's nit
```
rg -Io '\b(u?int\d+|size)_t\b' lib/ client-tools/ arangod/ | sort -n | uniq -c

     17 int16_t
    304 int32_t
    690 int64_t
     88 int8_t
   6423 size_t
    230 uint16_t
   1221 uint32_t
   4878 uint64_t
    743 uint8_t
```

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.h
Comment (MBkkt) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.h:50]: Why not? If you really don't like ifdef you can add single time

```cpp
#ifdef ARANGODB_USE_GOOGLE_TESTS
#define ADB_TEST_ONLY(expr) expr
#else
#define ADB_TEST_ONLY(expr)
#endif
```

File: arangod/Replication2/ReplicatedState/ReplicatedState.tpp
Comment (MBkkt) [reply arangod/Replication2/ReplicatedState/ReplicatedState.tpp:220]: Why don't you want use downCast if it's possible?

Comment (MBkkt) [reply arangod/Replication2/ReplicatedState/ReplicatedState.tpp:220]: The main reason is more clear code.
Now from code you cannot understand needs or not dynamic_cast

File: arangod/RocksDBEngine/RocksDBPersistedLog.cpp
Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.cpp:52]: Ok

Comment (MBkkt) [reply arangod/RocksDBEngine/RocksDBPersistedLog.cpp:52]: Ok, but in such case maybe add todo?

File: arangod/Cluster/UpdateReplicatedLogAction.cpp
Comment (MBkkt) [reply arangod/Cluster/UpdateReplicatedLogAction.cpp:67]: fixed

File: arangod/Replication2/Methods.cpp
Comment (MBkkt) [reply arangod/Replication2/Methods.cpp:806]: And fix it probably easier than writing answer to me :)

File: arangod/Replication2/ReplicatedState/ReplicatedState.tpp
Comment (MBkkt) [reply arangod/Replication2/ReplicatedState/ReplicatedState.tpp:220]: > Why don't you want use downCast if it's possible?

It also makes dynamic_cast check in maintainer mode
(if it's test only code, it's mean always)

But it also make static_cast which works only for base/derived classes

File: arangod/Replication2/ReplicatedLog/LogFollower.cpp
Comment (goedderz) [reply arangod/Replication2/ReplicatedLog/LogFollower.cpp]: Not since over a year, at least it's now called `checkResultAndCommitIndex`. But however its called, that lambda doesn't even capture the guard anymore since https://github.com/arangodb/arangodb/pull/16035, so the comment is somewhat obsolete.

File: arangod/VocBase/LogicalCollection.cpp
Comment (goedderz) [reply arangod/VocBase/LogicalCollection.cpp:1307]: The manager state in its previous form does not quite fit to the new state managers. However, I'd like to implement some pendant, because I assume it's useful for observability.

File: arangod/VocBase/LogicalCollection.cpp
Comment (goedderz) [reply arangod/VocBase/LogicalCollection.cpp:1266]: Note that `.get()` throws an exception if the result is an error.

More importantly, are we sure this can't happen? For example, can the shard (including the state) be deleted before this function is executed?

In which case we must not do any production assertion here, but handle it somewhat more gracefully.

File: tests/Replication2/Mocks/FakeReplicatedLog.h
Comment (goedderz) [reply tests/Replication2/Mocks/FakeReplicatedLog.h:158]: It's used in some of the tests that are currently disabled. I assume we won't need it in the future, because the tests will probably be rewritten or replaced without using it, but I'd rather keep it until then so it's easier to follow what the old tests did.

File: arangod/RocksDBEngine/RocksDBPersistedLog.cpp
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBPersistedLog.cpp:52]: I've opened a ticket. https://arangodb.atlassian.net/browse/CINFRA-648

File: arangod/Replication2/ReplicatedState/ReplicatedState.tpp
Comment (goedderz) [reply arangod/Replication2/ReplicatedState/ReplicatedState.tpp]: > The main reason is more clear code.
> Now from code you cannot understand needs or not dynamic_cast

I can get behind that.

File: lib/Basics/VelocyPackHelper.h
Comment (goedderz) [reply lib/Basics/VelocyPackHelper.h]: Maybe not as consistently, but it's used a lot in much of the AQL or Graph code, for example. But if you dislike it, I'll change it here.

File: arangod/Agency/Supervision.cpp
Comment (goedderz) [reply arangod/Agency/Supervision.cpp]: Changed it, I'll try to avoid mixing it in the future.

File: arangod/VocBase/vocbase.h
Comment (MBkkt) [reply arangod/VocBase/vocbase.h:204]: > I can still spot that will actually be noticeably is that we currently have no pretty printer for abseil, making it annoying to debug. 

Good point, will be fixed!

But as I wrote it's here not for making you immediately replace it, just for start discussion

File: arangod/Agency/Supervision.cpp
Comment (MBkkt) [reply arangod/Agency/Supervision.cpp]: Thanks!

File: lib/Basics/VelocyPackHelper.h
Comment (MBkkt) [reply lib/Basics/VelocyPackHelper.h]: Thanks!

File: arangod/VocBase/vocbase.h
Comment (MBkkt) [reply arangod/VocBase/vocbase.h:204]: Anyway I think it's need to be resolved.

**Review by MBkkt — APPROVED** [review]
Thanks for accepting suggestions!

In general I still think reserve before pushing will be definitely better, but it's replication2 code, so if you really don't want to do it, it's ok for me

File: arangod/VocBase/LogicalCollection.cpp
Comment (apetenchea) [reply arangod/VocBase/LogicalCollection.cpp:1266]: In `ReplicatedRocksDBTransactionCollection::ensureCollection` we end up asserting for the availability of the replicated state anyway. I think we're not supposed to reach that assert, but the current implementation is temporary anyway, and we'll have to change it once we introduce `distributeShardsLike`. I have nothing against using `.get()` here, as long as we'll be able where to look in case something blows up in a CI test.

File: arangod/Replication2/Methods.cpp
Comment (goedderz) [reply arangod/Replication2/Methods.cpp:806]: True. But you can turn that argument around as well. ;)

And of course you're right that a reserve is sensible from a performance perspective, I don't argue against that at all. But again, this is a debate of principles from my POV. The performance of these lines is just so way down on the list of priorities here that I don't bother. This code sends a network request to each participant, each of which possibly runs a log compaction, then waits for the answers. On the other hand, we're arguing about one or two additional (tiny) allocations. So I estimate we're talking about at least three, probably four, orders of magnitude difference.

File: arangod/VocBase/vocbase.h
Comment (goedderz) [reply arangod/VocBase/vocbase.h:204]: If you want to continue this discussion, shall we move it to DM?

File: arangod/VocBase/LogicalCollection.cpp
Comment (goedderz) [reply arangod/VocBase/LogicalCollection.cpp:1266]: Alright, thanks. I'll add an assertion for now, with a comment that we should check whether to remove it.

File: arangod/Replication2/Methods.cpp
Comment (MBkkt) [reply arangod/Replication2/Methods.cpp:806]: Yep

**Comment by goedderz** [comment]
> Thanks for accepting suggestions!
> 
> In general I still think reserve before pushing will be definitely better, but it's replication2 code, so if you really don't want to do it, it's ok for me

Thanks for taking the time to do a review!

**Comment by goedderz** [comment]
https://jenkins01.arangodb.biz/job/arangodb-matrix-pr/26673/
https://jenkins01.arangodb.biz/job/arangodb-ANY-linux-san.x86-64/1491/

**Pull request merged**
